middleware/proxy/dns.go

```diff
@@ -0,0 +1,78 @@
+package proxy
+
+import (
+	"net"
+
+	"github.com/miekg/coredns/middleware/pkg/singleflight"
+	"github.com/miekg/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+type dnsUpstream struct {
+	group *singleflight.Group
+}
+
+// ServeDNS does not satisfy middleware.Handler, instead it interacts with the upstream
+// and returns the respons or an error.
+func (d *dnsUpstream) Exchange(w dns.ResponseWriter, r *dns.Msg, address string) (*dns.Msg, error) {
```

Comment:
Do we need the address? Shouldn't the UpstreamHost already know that? In the code below you pass host.Name, which obviously it already knows.